build-support/scripts/test-flake.sh

```diff
@@ -0,0 +1,168 @@
+#!/bin/bash
+pushd $(dirname ${BASH_SOURCE[0]}) > /dev/null
+SCRIPT_DIR=$(pwd)
+pushd ../.. > /dev/null
+SOURCE_DIR=$(pwd)
+pushd build-support/docker > /dev/null
+IMG_DIR=$(pwd)
+popd > /dev/null
+
+source "${SCRIPT_DIR}/functions.sh"
+
+IMAGE="travis-img-v0.13"
+CONTAINER="travis-cnt"
+GOOS="linux"
+GOARCH="amd64"
+TEST_BINARY="flake.test"
+
+function usage {
+cat <<-EOF
+Usage: test-flake [<options ...>]
+
+Description:
+
+  test-flake surfaces flakiness in tests by constraining CPU resources.
+  
+  Single or package-wide tests are run for multiple iterations with a configurable
+  amount of CPU resources. 
+
+  0.15 CPUs and 30 iterations are configured as sane defaults.
+
+  See Docker docs for more info on tuning 'cpus' param: 
+  https://docs.docker.com/config/containers/resource_constraints/#cpu
+
+Options:
+
+  --pkg=""             Target package
+  --test=""            Target test (requires pkg flag)
+  --cpus=0.15          Amount of CPU resources for container
+  --n=30               Number of times to run tests
+
+Examples:
+
+  ./test-flake.sh --pkg connect/proxy
+  ./test-flake.sh --pkg connect/proxy --cpus 0.20
+  ./test-flake.sh --pkg connect/proxy --test Listener
+  ./test-flake.sh --pkg connect/proxy --test TestUpstreamListener
+  ./test-flake.sh --pkg connect/proxy --test TestUpstreamListener -n 100
+EOF
+}
+
+function build_repro_env {
+    # Arguments:
+    #   $1 - pkg, Target package
+    #   $2 - test, Target tests
+    #   $3 - cpus, Amount of CPU resources for container
+    #   $4 - n, Number of times to run tests
+
+    APP=$(pwd | awk '{n=split($0, a, "/"); print a[n]}')
+
+    echo -e "App:\t\t$APP"
+    echo -e "Package:\t$1"
+    echo -e "Test:\t\t$2"
+    echo -e "CPUs:\t\t$3"
+    echo -e "Iterations:\t$4"
+    echo
+
+    echo "-----> Cleaning up old containers..."
+    if docker ps -a | grep $CONTAINER ; then
+        docker rm $(docker ps -a | grep $CONTAINER | awk '{print $1;}')
+    fi
+
+    echo '----> Rebuilding image...'
+    (cd $IMG_DIR && docker build -q -t $IMAGE --no-cache -f Test-Flake.dockerfile .)
+
+    echo "---> Building app binary..."
+    env GOOS=$GOOS GOARCH=$GOARCH go build -o bin/$APP
+
+    echo "--> Building test binary..."
+    env GOOS=$GOOS GOARCH=$GOARCH go test -c "./$1" -o $TEST_BINARY
+
+    echo "-> Running container..."
+    docker run \
```

Comment:
What is the exit status of this if there is a test failure? Can it be used as the return code of the function?

Reply:
exit status is 1 when there's a failure, will also print the name of the test and the assertion output